### ISSUE TITLE
How does language versioning treat standalone scripts

### ISSUE THREAD

**Issue by leafpetersen** [dart-lang/language#771]
The language versioning feature is not entirely clear on what happens with standalone dart files that are not in a package.  The two relevant phrases that I can find are as follows:

```
The default language version of a Dart library is the default language version of the package that the library belongs to. If the library does not belong to a package, it has no default version.
```

and

```
Language versioning allows tools and compilers of an SDK to associate exactly one language version with each library in a program.

It does so by:

Assigning a default language version to each Dart package.
Assigning each Dart library to at most one Dart package.
Making each library inherit its package’s default language version unless it specifies a library-specific language version override.
Defaulting to the most recent language version if there is no other hint.
```

Is the intention that any standalone top level script is automatically opted into the new language feature?  This means that releasing a breaking language feature (like NNBD) will instantly break all standalone scripts.  Is this the intention, and is this really feasible?  

To avoid this, should we consider saying that standalone files (files not in a package) are defaulted to the lowest supported language version unless they are explicitly opted in?

cc @lrhn @munificent @vsmenon @jensjoha @stereotype441 @johnniwinther @jonasfj

**Comment by lrhn** [comment]
I see no good way around that.

A stand-alone script with no `pubspec.yaml`, no `package_config.json` and no version marker will be run at the most recent language level supported by the `dart` command. That's what you would want if you wrote the script from scratch right now.

The only other option I see is that absent of *any* configuration (meaning no packages available, no `package_config.json` found, no language marker), we limit the language version to the last version before language versioning was introduced.
That would mean that a *new* stand-alone script written a year from now, would need to add a `//@dart=2.10` to get the new semantics, otherwise it does not get NNBD enabled.

I have generally tried to make it so that getting the newest version of Dart is the default if you do nothing to prevent it, so that new users will get the newest language automatically.

I also think that scripts with no package dependencies are incredibly rare (so it's probably mostly myself getting hit by it :grin:).

**Comment by leafpetersen** [comment]
> I also think that scripts with no package dependencies are incredibly rare

Do we really believe this is true?  I'm very worried about, e.g. travis scripts, etc.

**Comment by lrhn** [comment]
I don't think Travis is an issue. It seems like it's all YAML configurations and Ruby and bash code which end up calling SDK programs like the analyzer, formatter or pub ("pub run test").

There might be other stand-alone programs. I know I have some in my `bin/` directory.
I do think those are rare because most other people use packages. 

If we support stand-alone scripts at all (and we should), then we should let people write a new script using the most recent language version. If we make `hello_world.dart` have different semantics than the same code put into a fresh package, or different from written on DartPad, then it's bad. I don't think locking a new script to an old language version is user friendly.

That does mean that an old script may break. So be it, the fix is easy (add `//@dart=2.7`) and the common use-case of having a pubspec.yaml still works. That's a one-time fix, and stand-alone-script authors who expect their script to endure will quickly learn to do that. People just writing `hello_world.dart` will see it working, and they won't come back to that code anyway.

I think this is basically down to making good error messages.
If your code compiles, we have to assume it's still correct. If it doesn't, then we can start making educated guesses about why. If the library has no version information available, and the error is consistent with the code being interpreted as the wrong version, we can suggest adding a version marker.

Any attempt to *guess* the version is going to be fragile. 
* Try parsing at version NNBD-1 first. If that succeeds (no `?` or `!` syntax) then assume it's pre-NNBD code. Can cause errors when it actually is NNBD-code that happens to be null-free.
* We could try compiling with the library at the newest version. If that fails, try again with the next newest version. Repeat as far back as we support. (Risks exponential blowup if we have multiple underspecified libraries)
* Use the "last modified" date of the file to guess a language version. Probably not stable or consistent across operating systems.

I think the simplest approach is the best: Treat the library as the newest version. There will be errors, but they will be very rare, very localized and easily fixable.

**Comment by leafpetersen** [comment]
Ok.  Let's run with that then.
